### PR TITLE
Change IAM binding to be authoritative

### PR DESCRIPTION
**Pull request by chrisst**
fixes https://github.com/terraform-providers/terraform-provider-google/issues/2449

-----------------------------------------------------------------
# [all]
Change IAM binding to be authoritative

### CODE REVIEW

File: third_party/terraform/resources/resource_iam_binding.go [hunk 1]
```diff
@@ -57,11 +57,8 @@ func resourceIamBindingCreate(newUpdaterFunc newResourceIamUpdaterFunc) schema.C
 
 		p := getResourceIamBinding(d)
 		err = iamPolicyReadModifyWrite(updater, func(ep *cloudresourcemanager.Policy) error {
-			// Creating a binding does not remove existing members if they are not in the provided members list.
-			// This prevents removing existing permission without the user's knowledge.
-			// Instead, a diff is shown in that case after creation. Subsequent calls to update will remove any
-			// existing members not present in the provided list.
-			ep.Bindings = mergeBindings(append(ep.Bindings, p))
+			// Creating a binding will overwrite the existing list of members for that role.
```
Comment (rileykarson) [inline third_party/terraform/resources/resource_iam_binding.go]: This doesn't seem surprising, especially with the line `overwriteBindings` below. Do you think we still need this comment?

File: third_party/terraform/resources/resource_iam_binding.go [hunk 2]
```diff
@@ -57,11 +57,8 @@ func resourceIamBindingCreate(newUpdaterFunc newResourceIamUpdaterFunc) schema.C
...
+			ep.Bindings = overwriteBinding(ep.Bindings, p)
```
Comment (rileykarson) [inline third_party/terraform/resources/resource_iam_binding.go]: `Update` is already authoritative; can we use a unified `CreateUpdate` method instead? We've done that in a few other places such as https://github.com/terraform-providers/terraform-provider-google/blob/master/google/resource_storage_default_object_acl.go#L11 for similar-ish resources.

File: third_party/terraform/utils/iam.go [hunk 3]
```diff
@@ -123,6 +123,25 @@ func iamPolicyReadModifyWrite(updater ResourceIamUpdater, modify iamPolicyModify
 	return nil
 }
 
+// Takes a single binding and will either overwrite the same role in a list or append it to the end
+func overwriteBinding(bindings []*cloudresourcemanager.Binding, overwrite *cloudresourcemanager.Binding) []*cloudresourcemanager.Binding {
```
Comment (rileykarson) [inline third_party/terraform/utils/iam.go]: Like above - this reads almost identically to `Update` in IAM Binding. A PR introducing this should include a reason why we couldn't use the exact same code for both.

File: third_party/terraform/website/docs/r/google_billing_account_iam_binding.md [hunk 4]
```diff
@@ -15,6 +15,9 @@ an existing Google Cloud Platform Billing Account.
    `google_billing_account_iam_member` for the __same role__ or they will fight over
    what your policy should be.
 
+~> **Note:** On create, this resource will overwrite members of any existing roles.
+    Use import if there are existing member that should be preserved.
```
Comment (rileykarson) [inline third_party/terraform/website/docs/r/google_billing_account_iam_binding.md]:
```suggestion
    Use `terraform import` and inspect the `terraform plan` output to ensure
    your existing members are preserved.
```

Minor wording nit(s)

File: third_party/terraform/website/docs/r/google_billing_account_iam_binding.md [hunk 5]
```diff
@@ -36,7 +39,7 @@ The following arguments are supported:
 
 * `role` - (Required) The role that should be applied.
 
-* `members` - (Required) A list of users that the role should apply to.
+* `members` - (Required) A list of users that the role should apply to. For more details on format and restrictions see https://cloud.google.com/billing/reference/rest/v1/Policy#Binding
```
Comment (rileykarson) [inline third_party/terraform/website/docs/r/google_billing_account_iam_binding.md]: This is a helpful change! These docs were very sparse. The same restrictions are true for `members` in all IAM resources afaik- is this worth bringing to `_member`, `_policy`, and `_audit_config`?

Comment (chrisst) [reply third_party/terraform/website/docs/r/google_billing_account_iam_binding.md]: Looks like _policy doesn't have members, but updated _member

File: third_party/terraform/utils/iam.go (hunk 3 above)
Comment (rileykarson) [inline third_party/terraform/utils/iam.go]: nit: I slightly prefer keeping this inlined in `CreateUpdate`

**Comment by rileykarson** [comment]
Sorry for the delay! I didn't notice the code pushes (I get emailed for comments but only a blue dot in GitHub for push)

File: third_party/terraform/website/docs/version_2_upgrade.html.markdown [hunk 6]
```diff
@@ -53,6 +53,7 @@ Upgrade topics:
 - [Resource: `google_sql_database_instance`](#resource-google_sql_database_instance)
 - [Resource: `google_storage_default_object_acl`](#resource-google_storage_default_object_acl)
 - [Resource: `google_storage_object_acl`](#resource-google_storage_object_acl)
+- [Resource: `google_*iam_binding`](#resource-TODO)
```
[the remark this answers is not in the document]
Comment (chrisst) [reply third_party/terraform/website/docs/version_2_upgrade.html.markdown]: ack

File: third_party/terraform/website/docs/version_2_upgrade.html.markdown [hunk 7]
```diff
@@ -623,3 +624,31 @@ values that were added outside of Terraform should be added to the config.
 Terraform will remove values not explicitly set in this field. Any `role_entity`
 values that were added outside of Terraform should be added to the config.
 For fine-grained management, use `google_storage_object_access_control`.
+
+## Resource: `google_*_iam_binding`
+
+### Create is now authoritative
+
+Every `iam_binding` resource will overwrite the existing member list for a given
+role on Create. Running `terraform plan` after create has succeeded will not
```
[the remark this answers is not in the document]
Comment (chrisst) [reply third_party/terraform/website/docs/version_2_upgrade.html.markdown]: re-worded.

File: third_party/terraform/website/docs/version_2_upgrade.html.markdown [hunk 8]
```diff
@@ -53,6 +53,7 @@ Upgrade topics:
 - [Resource: `google_sql_database_instance`](#resource-google_sql_database_instance)
 - [Resource: `google_storage_default_object_acl`](#resource-google_storage_default_object_acl)
 - [Resource: `google_storage_object_acl`](#resource-google_storage_object_acl)
+- [Resource: `google_*iam_binding`](#google_*_iam_binding)
```
Comment (rileykarson) [inline third_party/terraform/website/docs/version_2_upgrade.html.markdown]:
```suggestion
- [Resource: `google_*_iam_binding`](#google_*_iam_binding)
```

**Pull request merged**